### ISSUE TITLE
Compiler option for `!` to be compile-time only, no runtime check.

### ISSUE THREAD

**Issue by trusktr** [AssemblyScript/assemblyscript#1973]
From https://github.com/AssemblyScript/assemblyscript/issues/1972#issuecomment-878599169

**Comment by dcodeIO** [comment]
Can you provide some context? In particular, where would this be safe, since without a runtime check we'd lose type safety and as such would introduce potential hazards?

**Comment by jtenner** [comment]
Hey does this already work with --no-assert? 

Context: I've already done a pointer check and don't need to perform the actual assertion

**Comment by trusktr** [comment]
@willemneal What was the idea here? Maybe it is not possible to remove the runtime aspect?

**Comment by willemneal** [comment]
Consider:

```ts
function stringOrNull(s: string | null): string | null { return s; }

let s = stringOrNull("hello")!; // The type system should know this is non-null, but the `!` confirms it.
let n = stringOrNull(null)!; // Not safe
```

With the proposed option for `!` the first case above it is safe, while the second isn't.  However, this is also true for typescript.  The idea here is that by default `!` performs a runtime check to ensure that it's safe, but in the cases where the author knows better than the compiler it can help 

As for @jtenner's point, that would work, but it might be the case that the author wants bounds checks or other assertions but not this particular one.

**Comment by dcodeIO** [comment]
The difference to TS here is that if an author is wrong, they won't be greeted with a "Cannot do X on null" runtime error but with a potential hazard corrupting memory, leading to undefined behavior, a broken runtime or silent security issues.

**Comment by willemneal** [comment]
True, but it's not the default semantics and the author is responsible.

And it is possible currently to make the following function to achieve the same result.

```ts
@inline
function isNotNull<T>(t: T): NonNullable<T> { return t; }

let s = isNotNull(stringOrNull("hello")); 
// or
let s = <string>stringOrNull("hello")!;

```
 So really the change is a semantic sugar of the above approaches.

**Comment by dcodeIO** [comment]
Wait, are you saying that the use of `NonNullable<T>` is unsafe?

**Comment by willemneal** [comment]
Sorry it's
```ts
   function isNotNull<T>(t: T): NonNullable<T> { return <NonNullable<T>>t; }
```

and I updated my previous comment to remove the `!`.

And it's only unsafe if you use it in a situation where it could be null.

**Comment by dcodeIO** [comment]
That's similar to `<string>stringOrNull("hello")`. The cast there inserts a runtime check as well to be safe.

**Comment by willemneal** [comment]
Oh I didn't know that casting resulted in a runtime check.

Here is the `NonNullable` in safe place to allow for the next generic call to not have the null value in the type:
https://github.com/gagdiez/serial-as/blob/4a397fdd5e1bc6d303136f03912641cf00254d21/borsh/assembly/serializer.ts#L37

**Comment by dcodeIO** [comment]
In this case the compiler knows that `t` is never `null` and can omit the runtime check.

**Comment by willemneal** [comment]
Technically it shouldn't check in the first example since it should narrow the return type.

**Comment by dcodeIO** [comment]
TS doesn't do that, though:

```ts
function stringOrNull(s: string | null): string | null { return s; }
stringOrNull("asd").length; // Object is possibly 'null'.(2531)
```

**Comment by willemneal** [comment]
Ah shoot but it should!

I hate null and always have.  Perhaps we should look into multivalue return for a standard `Option` type.

**Comment by jtenner** [comment]
Yeah once multivalues are supported, Option would be amazing

**Comment by MaxGraey** [comment]
With option you got exactly the same. Just instead `str!`  you will got `str.unwrap()`, instead `str || ""` you will got `str.unwrap_or_else(() => "")` and etc.

**Comment by MaxGraey** [comment]
> Ah shoot but it should!

How? Same result in Rust:
```rust
fn string_or_null(s: Option<&str>) -> Option<&str> { 
  s
}

fn main() {
    let len = string_or_null(Some("hello")).len();
}
```
output:
```rust
  |
6 |     let len = string_or_null(Some("hello")).len();
  |                                             ^^^ method not found in `Option<&str>`

```

So you should do the same:
```rust
let len = string_or_null(Some("hello")).unwrap_or("").len();
```

**Comment by willemneal** [comment]
There is no null! And it will work for any type.  With multivalue we can also handle `Result` types so that `str.unwrap` would return the error. This style of error handling is great for AS since we don't have exceptions.

**Comment by willemneal** [comment]
> No, monadic Result and Option useful only in languages where it really builtin which also has pattern matching.

That's a bold statement.

**Comment by willemneal** [comment]
It'd still say it's useful even if it's more verbose.

**Comment by MaxGraey** [comment]
> It'd still say it's useful even if it's more verbose.

Why? Because it's done in Rust?

**Comment by MaxGraey** [comment]
Monadic Option and Result is very important concepts in pure functional languages because it's only one approach to handle and isolate side effects. But Rust is not Pure functional programming and even haven't builtin Monads. In other hand it haven't exceptions and nullability due to it's not fit into their paradigm. In my opinion monadic Option / Result in Rust look very unnatural and therefore verbose

**Comment by willemneal** [comment]
Rust also lets you do `stringOrNull(str)?.length`, which returns an result with an error.

**Comment by MaxGraey** [comment]
> Rust also lets you do `stringOrNull(str)?.length`, which returns an result with an error.

TypeScript also support this

**Comment by willemneal** [comment]
True, but it has exceptions so we'd need to either support that or use a result type.

**Comment by trusktr** [comment]
> True, but it has exceptions so we'd need to either support that or use a result type.

Or just replace anything that would return `undefined` with `null` and it would work in most cases. like `map.get('key that does not exist')` returns `null` instead of runtime error.

I _think_ I would be okay with `null` instead of errors where otherwise TS would have `undefined`. In my own code JS/TS, I always check `!foo` or `foo == null` or `foo != null`, and hence in these scenarios, the difference between `undefined` and `null` never matters. I think writing code any other way is asking for hazards.

So if we turns all `undefined` results into `null`, I think it would be good.

---

All of the following examples all work in TypeScript, but I have labeled them with `ERROR` or `OK` to show which ones work or don't work in AssemblyScript:

```ts
// All these should have the same final result as in TypeScript.
for (let node: Node | null = el.childNodes[0]!.firstChild; node; node = node.nextSibling) i++ // COMPILE ERROR
for (let node: Node | null = el.childNodes[0]!.firstChild; node; node = node!.nextSibling) i++ // OK
for (let node: Node | null = el.childNodes[0]!.firstChild; node!; node = node.nextSibling) i++ // COMPILE ERROR
for (let node: Node | null = el.childNodes[0]!.firstChild; node!; node = node!.nextSibling) i++ // RUNTIME ERROR
for (let node: Node | null = el.childNodes[0]!.firstChild; node != null; node = node.nextSibling) i++ // COMPILE ERROR
for (let node: Node | null = el.childNodes[0]!.firstChild; node != null; node = node!.nextSibling) i++ // OK
//
// These while-loop variants should all work the same too.
let node: Node | null = el.childNodes[0]!.firstChild
while (node) { node = node.nextSibling; i++ } // COMPILE ERROR
while (node) { node = node!.nextSibling; i++ } // OK
while (node!) { node = node.nextSibling; i++ } // COMPILE ERROR
while (node!) { node = node!.nextSibling; i++ } // RUNTIME ERROR
while (node != null) { node = node.nextSibling; i++ } // COMPILE ERROR
while (node != null) { node = node!.nextSibling; i++ } // OK
```

where `firstChild` and `nextSibling` are nullable types (`Node | null`).

**Comment by trusktr** [comment]
> See this example in [Hegel for example](https://hegel.js.org/try#GYVwdgxgLglg9mABAZygJxmA5geTQORABsiAKZASkQG9E0BTKENJZAbkQF8Aob1DbHkIlSAImRwAtvQDK6URQB0RetigALDgHotiOAGteKqCnQAuUwKyIAPojDEiiALz3HbPuky4Cj8uiUVNU1ERB1EejQ0ODQgA)

Yeah, Hegel is really nice in this regard.

**Comment by willemneal** [comment]
> Or just replace anything that would return undefined with null and it would work in most cases. like map.get('key that does not exist') returns null instead of runtime error.

Except for primitives.

**Comment by trusktr** [comment]
I forgot about that. That would be inconsistent (f.e. a Map with primitive values throws while a Map with object values returns nulls), but maybe it would be ok?

Would making primitives nullable incur runtime overhead? Or maybe it wouldn't have overhead once the type checker has better nullable awareness like Hegel's?

**Comment by MaxGraey** [comment]
> > See this example in [Hegel for example](https://hegel.js.org/try#GYVwdgxgLglg9mABAZygJxmA5geTQORABsiAKZASkQG9E0BTKENJZAbkQF8Aob1DbHkIlSAImRwAtvQDK6URQB0RetigALDgHotiOAGteKqCnQAuUwKyIAPojDEiiALz3HbPuky4Cj8uiUVNU1ERB1EejQ0ODQgA)
> 
> Yeah, Hegel is really nice in this regard. @JSMonk Can we convince you to join AssemblyScript efforts? 😃

Yes, we have been in the process of discussing the collaboration with @JSMonk for some time now. Stay tuned!

**Comment by willemneal** [comment]
If a primitive doesn't exist in a map then the type checker can't help, since it can't be known until runtime.  What I'm suggesting with the multivalue would solve it.  You would get back something like `(value, isNull)`, which would allow all types to be nullable.

**Comment by MaxGraey** [comment]
According exceptions vs Result monad. When you need return several error types simultaneously Result is really uncomfortable, non-performant (`Box<dyn error::Error>`) or verbose (a lot of boilerplates) in this case:
https://doc.rust-lang.org/rust-by-example/error/multiple_error_types.html

**Comment by willemneal** [comment]
`?` in TS only returns `undefined`, not return early.

```ts
function foo(val?: any): string? {
   let x = val?.func();
  return x?.charAt(0);
}
```

If `val` isn't defined or null, then `x` becomes undefined.

Whereas if it were used as in rust the function 
```ts
function foo(val: Option<any>): Result<string, Error> {
   let x = val?.func();
  return x.charAt(0);
}
```
would return with the error if `val` was `None`.

So the semantics compared to TS are different.  And I understand the difficulties in multiple error types, but it still seems the quickest path we have for error handling.

Perhaps the topic of error handling deserves a new issue or discussion.

**Comment by dcodeIO** [comment]
May I suggest to open a new issue, with concrete actionable items in its description, and a clear title?
